src/tracing/database.rs

```diff
@@ -0,0 +1,110 @@
+    Database, DatabaseCommit,
+};
+use reth_rpc_types::{serde_helpers::JsonStorageKey, BlockId, BlockNumberOrTag};
+use tokio::runtime::Handle;
+
+pub(crate) struct EthDatabaseSnapshot<P: EthereumProvider + Send + Sync> {
+    cache: CacheDB<P>,
+    block_id: BlockId,
+}
+
+impl<P: EthereumProvider + Send + Sync> EthDatabaseSnapshot<P> {
+    pub(crate) fn new(provider: P, block_id: BlockId) -> Self {
+        Self { cache: CacheDB::new(provider), block_id }
+    }
+}
+
+impl<P: EthereumProvider + Send + Sync> Database for EthDatabaseSnapshot<P> {
+    type Error = EthApiError;
+
+    fn basic(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        let cache = &self.cache;
+        if let Some(account) = cache.accounts.get(&address) {
+            return Ok(Some(account.info.clone()));
+        }
+
+        let account_info = Handle::current().block_on(async {
+            let bytecode = cache.db.get_code(address, Some(self.block_id)).await?;
+            let bytecode = Bytecode::new_raw(bytecode);
+            let code_hash = bytecode.hash_slow();
+
+            let nonce = cache.db.transaction_count(address, Some(self.block_id)).await?.to();
+            let balance = cache.db.balance(address, Some(self.block_id)).await?;
+
+            Result::<_, EthApiError>::Ok(AccountInfo { nonce, balance, code: Some(bytecode), code_hash })
+        })?;
+
+        self.cache.insert_account_info(address, account_info.clone());
+        Ok(Some(account_info))
+    }
+
+    fn code_by_hash(&mut self, code_hash: B256) -> Result<Bytecode, Self::Error> {
+        if let Some(bytecode) = self.cache.contracts.get(&code_hash) {
+            return Ok(bytecode.clone());
+        }
+        Ok(Bytecode::default())
+    }
+
+    fn storage(&mut self, address: Address, index: U256) -> Result<U256, Self::Error> {
+        let cache = &self.cache;
+        if let Some(account) = cache.accounts.get(&address) {
+            if let Some(storage) = account.storage.get(&index) {
+                return Ok(*storage);
+            }
+        }
+
+        let storage = Handle::current().block_on(async {
+            let value = cache
+                .db
+                .storage_at(address, JsonStorageKey(B256::from_slice(&index.to_be_bytes::<32>())), Some(self.block_id))
+                .await?;
+            Result::<_, EthApiError>::Ok(value)
+        })?;
+        let storage = U256::from_be_bytes(storage.0);
+
+        self.cache.accounts.entry(address).or_default().storage.insert(index, storage);
+        Ok(storage)
+    }
+
+    fn block_hash(&mut self, number: U256) -> Result<B256, Self::Error> {
+        let cache = &self.cache;
+        if let Some(hash) = cache.block_hashes.get(&number) {
+            return Ok(*hash);
+        }
+
+        let hash = Handle::current().block_on(async {
+            let hash = cache
+                .db
+                .block_by_number(BlockNumberOrTag::Number(number.to()), false)
```

Comment:
shouldn't we put a `try_into` with error handling here as `number` is `U256` and the required type for `BlockNumberOrTag::Number` is `u64`?